profiler/src/ProfilerEngine/Datadog.Profiler.Native/RuntimeIdStore.cpp

```diff
@@ -0,0 +1,203 @@
+    _getIdFn = reinterpret_cast<const std::string&(STDMETHODCALLTYPE*)(AppDomainID)>(externalFunction);
+    return _getIdFn != nullptr;
+#endif
+}
+
+bool RuntimeIdStore::Stop()
+{
+#ifdef LINUX
+    return true;
+#else
+    if (_instance != nullptr)
+    {
+        return FreeDynamicLibrary(_instance);
+    }
+
+    Log::Warn("RuntimeID store service was not started correctly.");
+    return false;
+#endif
+}
+
+const char* RuntimeIdStore::GetName()
+{
+    return RuntimeIdStore::ServiceName;
+}
+
+const std::string& RuntimeIdStore::GetId(AppDomainID appDomainId)
+{
+    if (_getIdFn != nullptr)
+    {
+        return _getIdFn(appDomainId);
+    }
+
+    auto& rid = _runtimeIdPerAppdomain[appDomainId];
+
+    if (rid.empty())
+    {
+        rid = GenerateRuntimeId();
+    }
+
+    return rid;
+}
+
+void* RuntimeIdStore::LoadDynamicLibrary(std::string filePath)
+{
+    Log::Debug("LoadDynamicLibrary: ", filePath);
+
+#if _WIN32
+    HMODULE dynLibPtr = LoadLibrary(::shared::ToWSTRING(filePath).c_str());
+    if (dynLibPtr == NULL)
+    {
+        LPVOID msgBuffer;
+        DWORD errorCode = GetLastError();
+
+        FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+                      NULL, errorCode, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), (LPTSTR)&msgBuffer, 0, NULL);
+
+        if (msgBuffer != NULL)
+        {
+            Log::Warn("LoadDynamicLibrary: Error loading dynamic library '", filePath, "': ", (LPTSTR)msgBuffer);
+            LocalFree(msgBuffer);
+        }
+    }
+    return dynLibPtr;
+#else
+    void* dynLibPtr = dlopen(filePath.c_str(), RTLD_LOCAL | RTLD_LAZY);
+    if (dynLibPtr == nullptr)
+    {
+        char* errorMessage = dlerror();
+        Log::Warn("LoadDynamicLibrary: Error loading dynamic library '", filePath, "': ", errorMessage);
+    }
+    return dynLibPtr;
+#endif
+}
+
+void* RuntimeIdStore::GetExternalFunction(void* instance, const char* const funcName)
+{
+    Log::Debug("GetExternalFunction: ", funcName);
+
+    if (instance == nullptr)
+    {
+        Log::Warn("GetExternalFunction: The module instance is null.");
+        return nullptr;
+    }
+
+#if _WIN32
+    FARPROC dynFunc = GetProcAddress((HMODULE)instance, funcName);
+    if (dynFunc == NULL)
+    {
+        LPVOID msgBuffer;
+        DWORD errorCode = GetLastError();
+
+        FormatMessage(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_SYSTEM | FORMAT_MESSAGE_IGNORE_INSERTS,
+                      NULL, errorCode, MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT), (LPTSTR)&msgBuffer, 0, NULL);
+
+        if (msgBuffer != NULL)
+        {
+            Log::Warn("GetExternalFunction: Error loading dynamic function '", funcName, "': ", (LPTSTR)msgBuffer);
+            LocalFree(msgBuffer);
+        }
+    }
+    return dynFunc;
+#else
+    void* dynFunc = dlsym(instance, funcName);
+    if (dynFunc == nullptr)
+    {
+        char* errorMessage = dlerror();
```

Comment:
is it needed to "free" that string?

Reply:
this is what''s  done elsewhere. I'll check if we need or not to free it.

Reply:
nop, it a pointer returned by `dlerror` and reused.